Work around CF timers not being serviced promptly

bnham opened the pull request:
#### 7b65bcf29a4d2c738d2dfa89f05c9d48bde0237f
<pre>
Work around CF timers not being serviced promptly
<a href="https://bugs.webkit.org/show_bug.cgi?id=301665">https://bugs.webkit.org/show_bug.cgi?id=301665</a>
<a href="https://rdar.apple.com/154763428">rdar://154763428</a>

Reviewed by Brady Eidson.

We have a long-standing bug where WorkerDedicatedRunLoop::runInMode sometimes consumes 100% of a
CPU. After adding logging for users running in to this, we&apos;ve discovered this is because CFRunLoop
gets in to a state where CFRunLoopGetNextTimerFireDate returns a fire date in the distant past (say
~10 minutes ago), but the actual next alarm wakeup for the run loop is in the distant future (say ~5
minutes in the future). Since we&apos;re running CFRunLoop in polling mode (timeout=0), this causes us to
spin the CPU at 100% calling CFRunLoopRunInMode until the next alarm goes off (which would be 5
minutes of CPU spin in this example).

This is actually expected because while CFRunLoopTimerSetNextFireDate takes a wall clock time (which
does advance while the system sleeps), it actually creates a Mach timer based on mach_absolute_time
(which doesn&apos;t advance when the system sleeps). So if the system goes to sleep while a CF timer is
armed, then after the system resumes CFRunLoopGetNextTimerFireDate may return a wall timestamp in
the past, even though the timer will actually fire in the future.

There doesn&apos;t seem to be any correct way of fixing this other than rewriting our run loop to have
CFRunLoop drive everything. The current implementation that tries to ping pong between our run loop
and CFRunLoop while using CFRunLoopGetNextTimerFireDate to figure out if we have to spin the
CFRunLoop doesn&apos;t work given the issue I just described and can cause the CPU to spin until the
run loop timer actually fires.

This commit adds a mitigation that we can ship until we can refactor WorkerDedicatedRunLoop.
Basically, when we detect that there&apos;s a timer that should have fired in the distant past (more than
one second ago), we have CFRunLoop block for up to 1 second. In the buggy case that we&apos;re trying to
mitigate, this will cause us to wake up at most once per second, which is less than optimal but a
lot better than an infinite loop. In the non-buggy case, we don&apos;t end up with any extra perf penalty
because we now instruct CFRunLoop to exit immediately after it handles a source (which it will in
the normal case that the next timer fire date is in the past and the Mach timer did actually fire in
the past).

* Source/WebCore/workers/WorkerRunLoop.cpp:
(WebCore::WorkerDedicatedRunLoop::runInMode):

Canonical link: <a href="https://commits.webkit.org/302518@main">https://commits.webkit.org/302518@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8823c1bfbb7e4e5ce2eb50a7b642542218bdc4cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1584 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40166 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136705 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80724 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/26672865-a871-4b8f-9c1e-1ba8c2069ac7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131199 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1461 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98494 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66394 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/34e58f36-8e16-408e-97e6-7cb76eb99668) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1182 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115838 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79144 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/feeeb463-b283-4bc1-9672-ecfce4354f8d) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33966 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79982 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34467 "Found 2 new test failures: fast/images/page-wide-animation-toggle.html http/tests/site-isolation/window-open-with-name-cross-site.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139178 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1375 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107021 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1417 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106864 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27211 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/1122 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30701 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54011 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1446 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64809 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1263 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1299 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1368 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/workers/WorkerRunLoop.cpp:
```diff
@@ -284,7 +284,7 @@ WorkerDedicatedRunLoop::RunInModeResult WorkerDedicatedRunLoop::runInMode(Worker
     Seconds timeoutDelay = Seconds::infinity();
 
 #if USE(CF)
-    CFAbsoluteTime nextCFRunLoopTimerFireDate = CFRunLoopGetNextTimerFireDate(RetainPtr { CFRunLoopGetCurrent() }.get(), kCFRunLoopDefaultMode);
```
We should probably wrap CFRunLoopGetCurrent() in a function which ends with "Singleton".
 Otherwise, safer C++ bot will complain.

bnham replied:
Will fix

bnham committed:
Work around CF timers not being serviced promptly
https://bugs.webkit.org/show_bug.cgi?id=301665
rdar://154763428

Reviewed by Brady Eidson.

We have a long-standing bug where WorkerDedicatedRunLoop::runInMode sometimes consumes 100% of a
CPU. After adding logging for users running in to this, we've discovered this is because CFRunLoop
gets in to a state where CFRunLoopGetNextTimerFireDate returns a fire date in the distant past (say
~10 minutes ago), but the actual next alarm wakeup for the run loop is in the distant future (say ~5
minutes in the future). Since we're running CFRunLoop in polling mode (timeout=0), this causes us to
spin the CPU at 100% calling CFRunLoopRunInMode until the next alarm goes off (which would be 5
minutes of CPU spin in this example).

This is actually expected because while CFRunLoopTimerSetNextFireDate takes a wall clock time (which
does advance while the system sleeps), it actually creates a Mach timer based on mach_absolute_time
(which doesn't advance when the system sleeps). So if the system goes to sleep while a CF timer is
armed, then after the system resumes CFRunLoopGetNextTimerFireDate may return a wall timestamp in
the past, even though the timer will actually fire in the future.

There doesn't seem to be any correct way of fixing this other than rewriting our run loop to have
CFRunLoop drive everything. The current implementation that tries to ping pong between our run loop
and CFRunLoop while using CFRunLoopGetNextTimerFireDate to figure out if we have to spin the
CFRunLoop doesn't work given the issue I just described and can cause the CPU to spin until the
run loop timer actually fires.

This commit adds a mitigation that we can ship until we can refactor WorkerDedicatedRunLoop.
Basically, when we detect that there's a timer that should have fired in the distant past (more than
one second ago), we have CFRunLoop block for up to 1 second. In the buggy case that we're trying to
mitigate, this will cause us to wake up at most once per second, which is less than optimal but a
lot better than an infinite loop. In the non-buggy case, we don't end up with any extra perf penalty
because we now instruct CFRunLoop to exit immediately after it handles a source (which it will in
the normal case that the next timer fire date is in the past and the Mach timer did actually fire in
the past).

* Source/WebCore/workers/WorkerRunLoop.cpp:
(WebCore::WorkerDedicatedRunLoop::runInMode):

Canonical link: https://commits.webkit.org/302518@main
